src/runtime-rs/crates/hypervisor/src/qemu/network.rs

```diff
@@ -0,0 +1,372 @@
+    type Error = anyhow::Error;
+
+    // method to parse a PciPath from a string
+    fn try_from(s: &str) -> Result<Self> {
+        let transport = match s {
+            "virtio-net-pci" => NetDeviceWithTransport::VirtioNetPci,
+            "virtio-net-ccw" => NetDeviceWithTransport::VirtioNetCcw,
+            "virtio-net" => NetDeviceWithTransport::VhostUserNet,
+            "vfio-pci" => NetDeviceWithTransport::VfioPci,
+            "vfio-ccw" => NetDeviceWithTransport::VfioCcw,
+            "vfio-ap" => NetDeviceWithTransport::VfioAp,
+            _ => return Err(anyhow!("unsupported transport")),
+        };
+
+        Ok(transport)
+    }
+}
+
+#[derive(Debug, Default, PartialEq)]
+enum NetDeviceDriver {
+    #[default]
+    VirtioNet,
+    Vhostuser,
+    Vfio,
+}
+
+impl ToString for NetDeviceDriver {
+    fn to_string(&self) -> String {
+        match self {
+            NetDeviceDriver::VirtioNet => "virtio-net".to_owned(),
+            NetDeviceDriver::Vhostuser => "vhost-user".to_owned(),
+            NetDeviceDriver::Vfio => "vfio".to_owned(),
+        }
+    }
+}
+
+#[derive(Debug, Default, PartialEq)]
+enum NetDeviceType {
+    /// Tap is a tap networking device type.
+    /// TAP NetDeviceType = "tap"
+    #[default]
+    Tap,
+    /// MacTap is a macvtap networking device type.
+    /// MACVTAP NetDeviceType = "macvtap"
+    MacvTap,
+    /// IpTap is a ipvtap virtual networking device type.
+    /// IPVTAP NetDeviceType = "ipvtap"
+    IpvTap,
+    /// VethTap is a veth-tap virtual networking device type.
+    /// VETHTAP NetDeviceType = "vethtap"
+    VethTap,
+    /// VhuNet is a vhost-user port (socket)
+    /// VhuNet NetDeviceType = "vhostuser"
+    VhuNet,
+    /// Physical is a direct assigned PCI device or PCI VF
+    /// Physical NetDeviceType = "VFIO"
+    Physical,
+}
+
+impl NetDeviceType {
+    // QemuNetdevParam converts to the QEMU -netdev parameter notation
+    fn qemu_netdev_param(&self, transport: VirtioTransport) -> Option<&str> {
+        match self {
+            // -netdev type=tap -device virtio-net-pci
+            NetDeviceType::VethTap
+            | NetDeviceType::Tap
+            | NetDeviceType::MacvTap
+            | NetDeviceType::IpvTap => Some("tap"),
+            // -netdev type=vhost-user (no device)
+            NetDeviceType::VhuNet => {
+                if transport == VirtioTransport::Ccw {
+                    warn!(sl!(), "vhost-user devices are not supported on IBM Z");
+                    return None;
+                }
+                Some("vhost-user")
+            }
+            // -device vfio-pci (no netdev)
+            NetDeviceType::Physical => {
+                if transport == VirtioTransport::Mmio {
+                    warn!(
+                        sl!(),
+                        "vfio devices are not support with the MMIO transport"
+                    );
+                    return None;
+                }
+                Some("")
```

Comment:
As mentioned above, `-device vfio` comes with no `-netdev`. It seems that this function should not even be called in the vfio case.